docs/references/http-gateway-protocol-spec.md

```diff
@@ -0,0 +1,471 @@
+
+## Upgrade to Update Calls
+
+If the canister sets `upgrade = opt true` in the `HttpResponse` reply from `http_request`, then the Gateway ignores all other fields of the response. The Gateway performs an [update](https://internetcomputer.org/docs/current/references/ic-interface-spec#http-call) call to `http_request_update`, passing the same `HttpRequest` record as the argument, and uses the resulting response from `http_request_update` instead.
+
+The value of the `upgrade` field returned from `http_request_update` is ignored.
+
+## Legacy Response Verification
+
+Version 1 response verification only supports verifying a request path and response body pair with only one response per request path. This is quite restrictive in the number of scenarios it can support. For example, redirection or client-side caching is not safe since the status code and headers required to verify responses of that nature are not included in the certification. Upon a query call to a canister’s `http_request` method, a single malicious node or boundary node can modify these parts of the HTTP response, leading to the following issues:
+
+- For any dApp, a single malicious node on the subnet can maliciously redirect the user anywhere they like.
+- dApps cannot load the service worker when embedded within iFrames.
+- The use of redirects and cookies is unsafe as they can be manipulated by malicious nodes.
+- This is unexpected for developers and will lead to vulnerabilities in dApps sooner or later.
+- The effectiveness of security headers (such as Content Security Policy) is diminished as they can be omitted or modified by malicious nodes.
+
+[Response Verification version 2](#response-verification) overcomes these issues.
+
+The steps for response verification are as follows:
+
+- See the [response verification outline](#response-verification-outline) for the full subprotocol description.
+- The path `["http_assets", <url>]` exists in the `tree` and is a leaf with a value, where `<url>` is the utf8-encoded URL from the `HttpRequest`.
+- Otherwise, the path `["http_assets", "/index.html"]` must exist in the `tree` and be a leaf.
+- That leaf must contain the SHA-256 hash of the decoded body.
+  - If the `streaming_strategy` field of the `HttpResponse` is set, all chunks are streamed and concatenated according to [response body streaming](#response-body-streaming) before decoding.
+  - The body is decoded according to the `Content-Encoding` header if present.
+  - Supported values for the `Content-Encoding` header include `gzip` and `deflate`.
+
+## Canister HTTP Interface
+
+The full [Candid](https://github.com/dfinity/candid/blob/master/spec/Candid.md) interface that a canister is expected to implement is as follows:
+
+```
+type HeaderField = record { text; text; };
```

Comment:
Again the same thing, but I really have no clue of candid:
```suggestion
type HeaderField = record { text: text; };
```

Reply:
`record { text: text; };` will create a record that has a single entry with the literal label `"text"` and a value of type `"text"`.
But `record { text; text}` creates a record with two entries that have auto-generated labels and values of type `"text"`, so it's the equivalent of writing `record { 0: text, 1: text }`: https://internetcomputer.org/docs/current/references/candid-ref#type-record--n--t--

Reply:
Thanks for the explanation! Makes sense :)